Library/OcPciIoLib/OcPciIoU.c

```diff
@@ -0,0 +1,446 @@
+   EfiCpuIoWidthUint64 do not work. Misaligned buffers, on the other hand, will
+   be handled by the driver.
+
+   If Width is EfiCpuIoWidthUint8, EfiCpuIoWidthUint16, EfiCpuIoWidthUint32,
+   or EfiCpuIoWidthUint64, then both Address and Buffer are incremented for
+   each of the Count operations that is performed.
+
+   If Width is EfiCpuIoWidthFifoUint8, EfiCpuIoWidthFifoUint16,
+   EfiCpuIoWidthFifoUint32, or EfiCpuIoWidthFifoUint64, then only Buffer is
+   incremented for each of the Count operations that is performed. The read or
+   write operation is performed Count times on the same Address.
+
+   If Width is EfiCpuIoWidthFillUint8, EfiCpuIoWidthFillUint16,
+   EfiCpuIoWidthFillUint32, or EfiCpuIoWidthFillUint64, then only Address is
+   incremented for each of the Count operations that is performed. The read or
+   write operation is performed Count times from the first element of Buffer.
+
+   @param[in]  This     A pointer to the EFI_CPU_IO2_PROTOCOL instance.
+   @param[in]  Width    Signifies the width of the I/O or Memory operation.
+   @param[in]  Address  The base address of the I/O operation.
+   @param[in]  Count    The number of I/O operations to perform. The number of
+                       bytes moved is Width size * Count, starting at Address.
+   @param[in]  Buffer   For read operations, the destination buffer to store the results.
+                       For write operations, the source buffer from which to write data.
+
+   @retval EFI_SUCCESS            The data was read from or written to the PI system.
+   @retval EFI_INVALID_PARAMETER  Width is invalid for this PI system.
+   @retval EFI_INVALID_PARAMETER  Buffer is NULL.
+   @retval EFI_UNSUPPORTED        The Buffer is not aligned for the given Width.
+   @retval EFI_UNSUPPORTED        The address range specified by Address, Width,
+                                 and Count is not valid for this PI system.
+
+ **/
+EFI_STATUS
+EFIAPI
+CpuMemoryServiceWrite (
+  IN EFI_CPU_IO2_PROTOCOL       *This,
+  IN EFI_CPU_IO_PROTOCOL_WIDTH  Width,
+  IN UINT64                     Address,
+  IN UINTN                      Count,
+  IN VOID                       *Buffer
+  )
+{
+  EFI_STATUS                 Status;
+  UINT8                      InStride;
+  UINT8                      OutStride;
+  EFI_CPU_IO_PROTOCOL_WIDTH  OperationWidth;
+  UINT8                      *Uint8Buffer;
+
+  Status = CpuIoCheckParameter (TRUE, Width, Address, Count, Buffer);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  //
+  // Select loop based on the width of the transfer
+  //
+  InStride       = mInStride[Width];
+  OutStride      = mOutStride[Width];
+  OperationWidth = (EFI_CPU_IO_PROTOCOL_WIDTH)(Width & 0x03);
+  for (Uint8Buffer = Buffer; Count > 0; Address += InStride, Uint8Buffer += OutStride, Count--) {
+    if (OperationWidth == EfiCpuIoWidthUint8) {
+      MmioWrite8 ((UINTN)Address, *Uint8Buffer);
```

Comment:
Use ReadUnalignedXX functions here and below.

Reply:
Not resolved, use direct access